### ISSUE TITLE
Add Multiple File Upload component

### ISSUE THREAD

**Issue by mcarrano** [patternfly/patternfly-react#6683]
Adds a new component that supports multiple file upload. 

Design: https://marvelapp.com/prototype/69jc198
Core PR: https://github.com/patternfly/patternfly/pull/4548

**Comment by mcarrano** [comment]
@jessiehuff can you take a look from an a11y perspective? Thanks.

**Comment by megan-hall** [comment]
Just as a note - the component should support the ability to upload a single file, restrict file type and file size as necessary.

**Comment by wise-king-sullyman** [comment]
What is the loading bar intended to really show? Should it be showing how much of each file has been loaded into memory by the component (or something else internal to the component?), or should it be something that the consumers interact with directly so that wherever the files are being uploaded *to* can tell the component what amount of the file they've received at any point in time?

**Comment by mcarrano** [comment]
That's an interesting question @wise-king-sullyman . @megan-hall do you have any example use cases for how this component might be used? I guess the answer depends on where the component is uploading the file to. Is it just pulling the file into browser memory or does the consumer specify where the files are going? I think in general we should make things as flexible as possible to allow consumers control.

**Comment by megan-hall** [comment]
In most use cases that I am aware of, the user is just fulling the filed into browser memory and not specifying where the file is uploading to. Ex. attaching a image to a support case.

**Comment by wise-king-sullyman** [comment]
@megan-hall so to confirm, the amount of the file that's been read into memory is what you would like to see displayed by the loading bar, and that should be done inside the component?

Or would it be better to try and have the component expose the value loaded in some way; so that what the loading bar shows is up to the consumer, and include an example (or demo) that demonstrates how to pass in the amount loaded?

**Comment by megan-hall** [comment]
Do we have a best practices approach for PF in terms leaving this up to the consumer? Or handling inside the component? (this is where I begin to default to Matt).

**Comment by mcarrano** [comment]
Yeah this is where I usually defer to @nicolethoen or @tlabaj for the right answer. @mturley I know that you worked on the simple File Upload so perhaps you have an opinion on this?

**Comment by mturley** [comment]
I would argue that we definitely need a way for the consumer to control the loading bar, even if we have default behavior that doesn't leave it entirely up to the consumer. File upload/import can be handled a number of different ways by the consumer and we don't want to restrict the component too much.

For the default behavior of the single-file upload component, it has to load the file contents synchronously which blocks the main thread, and I had trouble trying to get it to even render a spinner during that process. There may be a better way to implement that and drive the progress bar that way, but if it's a large enough file that takes significant time to load, the consumer is better off handling the upload themselves anyway so they can skip having to cache it in browser memory. Certain very large files can't be cached that way anyway. So the single-file upload component allows the consumer to custom-render it however they need it to look and drive the file handling entirely themselves.

If we want to try and find a good way to render the progress bar while the component loads the file into memory as a default mode we can do that, but I think we definitely need to allow the consumer to control the progress bar so they can use it to reflect a separate network upload instead if that's what they need.

**Comment by wise-king-sullyman** [comment]
@mcarrano @megan-hall should this component prevent users from uploading identical files multiple times?

If so what do you think that behavior should look like?

**Comment by mcarrano** [comment]
@wise-king-sullyman This is a great question. @megan-hall is on PTO for a couple of days, so I'll attempt to answer. Seems like we should prevent this. Do you agree? If so, then I think that the error would be flagged as soon as you try to drop or pick the file before starting to upload. Unfortunately we didn't define error styles for that (we only have the provision for displaying an error that occurs during upload). But if that's how we want to go, then I think we can talk about how this will display and potentially will need a quick core update to add styling.

**Comment by wise-king-sullyman** [comment]
@mcarrano I agree that it should probably be prevented. I can't think of a reason why a user would need to upload multiple copies of the same file and enabling that functionality would increase the complexity of this component.

**Comment by mcarrano** [comment]
@wise-king-sullyman Do you want to go ahead and code it that way and perhaps once @megan-hall returns we can figure out what that error message should look like?

**Comment by megan-hall** [comment]
@wise-king-sullyman @mcarrano I was looking at how google drive handles these interactions. 

For both drag and drop and file selection they surface the modal below. We could do something similar but if the "keep both files" adds complexity perhaps that is something we could consider adding in the future if necessary.

<img width="537" alt="Screen Shot 2022-02-03 at 10 33 23 AM" src="https://user-images.githubusercontent.com/62031580/152375058-388b2a46-a3c1-4068-952b-824109233917.png">

**Comment by wise-king-sullyman** [comment]
@megan-hall can you think of a use case for a user needing to upload multiple copies of the same file?

I'd be fine with trying to implement something like that, I'm just not sure of the utility. If we are going to do a modal with an option to keep both somehow I *would* like if that could be done in the future under a separate milestone though.

**Comment by megan-hall** [comment]
@wise-king-sullyman Off of the top of my head I don't know of a particular use case to support multiple copies of the same file.

**Comment by mcarrano** [comment]
@wise-king-sullyman @megan-hall I think we might be talking about two different use cases that are related but not the same. In the first case, the user is attempting to upload the same file twice in the same session, e.g. maybe they selected the wrong file in error or want to retry the upload. I think that unless the upload failed the first time, the component should not attempt to upload the same file again and an error should be returned. I guess that the design question is what that error should look like. @megan-hall I'd be glad to get together and brainstorm on that a bit. @wise-king-sullyman do you agree that it makes sense to allow the user to retry uploading the same file again if an initial upload fails?

The second case, as illustrated by the Google example, is what to do if the user attempts to upload a file that already exists at the destination. I agree that we don't want to hardwire a modal or any other specific handling for that case into the component, but it would be good if that could be passed off to the consumer to write their own handler. Does that make sense @wise-king-sullyman ?

**Comment by wise-king-sullyman** [comment]
@mcarrano yes I totally agree that if the upload fails the file should be re-uploadable again, and I imagine the failed attempt should just be thrown out when the re-upload is attempted, do you agree with that?

Yes letting the consumer handle that error makes sense to me.

**Comment by mcarrano** [comment]
@mcarrano yes I totally agree that if the upload fails the file should be re-uploadable again, and I imagine the failed attempt should just be thrown out when the re-upload is attempted, do you agree with that?

Yes. Agree.

**Comment by mcarrano** [comment]
@wise-king-sullyman I just met with @megan-hall to discuss what to do about the duplicate file case. We realized that there could be legitimate reason for repeating the upload of a file with the same name. For example this could occur if the user selected the first file from the wrong location (i.e. a different folder) or if the user forgot to make an edit and re-saves the file with the same name. We therefore thought that maybe we should let the file be uploaded again which would replace the first version. There seems to be no harm in this except wasting some time to allow the upload to take place a second time. Do you agree? This would avoid us needing to create a new error state that was not defined in the initial design.

If we do this, it should be the default behavior. However the consumer should still have the ability to substitute their own error handling which could include attaching a modal or similar mechanism to give the user more control. Let us know your thoughts.

**Comment by wise-king-sullyman** [comment]
@mcarrano @megan-hall Currently I'm evaluating "sameness" of two files using the combination of their name, file size, and time of late modification; but if you'd like me to rework things to only go off of the name and just trigger a re-upload whenever a file with the same name is used I can do that. I don't think  that either approach is really *wrong* and I'm happy to go in either direction that you all think is best.

I agree that we should ultimately leave this sort of decision / general error handling up to the consumers and I've taken a very composable approach to the creation of this component to try and keep things as customizable by the consumer as possible. I've kept as much logic as possible outside of the components themselves. It's really just a matter of what functionality we want to be demonstrated to the consumers via the examples.

**Comment by mcarrano** [comment]
@wise-king-sullyman I don't feel strongly either way, either. Sounds like your planned duplicate file detection is pretty robust and would allow for the cases we were considering. I still think letting the upload proceed by default but allowing the consumer to wire in more complex error handling is the way to go. What do you think @megan-hall ?

**Comment by mturley** [comment]
I wonder if any built-in logic that makes assumptions about the nature of files like duplicate detection should have an optional override for the consumer to pass in, e.g. a function passed as a `compareFiles` prop that is used if present?

**Comment by wise-king-sullyman** [comment]
@mturley I just finished a rework of the component to further increase composability that will have any duplicate detection or handling occur on the consumer side.

Pretty much the only logic I have built-in at the moment is in the file upload status item component which shows the details and loading status/result for each individual file. I am providing props on it to allow consumers to bypass the built-in logic and just use it as a display component if desired though, as you're suggesting here.
